contractcourt/htlc_timeout_resolver.go

```diff
@@ -1314,3 +1227,92 @@ func (h *htlcTimeoutResolver) sweepTimeoutTxOutput() error {
 
 	return err
 }
+
+// checkpointStageOne creates a checkpoint for the first stage of the htlc
+// timeout transaction. This is used to ensure that the resolver can resume
+// watching for the second stage spend in case of a restart.
+func (h *htlcTimeoutResolver) checkpointStageOne(
+	spendTxid chainhash.Hash) error {
+
+	h.log.Debugf("checkpoint stage one spend of HTLC output %v, spent "+
+		"in tx %v", h.outpoint(), spendTxid)
+
+	// Now that the second-level transaction has confirmed, we checkpoint
+	// the state so we'll go to the next stage in case of restarts.
+	h.outputIncubating = true
+
+	// Create stage-one report.
+	report := &channeldb.ResolverReport{
+		OutPoint:        h.outpoint(),
+		Amount:          h.htlc.Amt.ToSatoshis(),
+		ResolverType:    channeldb.ResolverTypeOutgoingHtlc,
+		ResolverOutcome: channeldb.ResolverOutcomeFirstStage,
+		SpendTxID:       &spendTxid,
+	}
+
+	// At this point, the second-level transaction is sufficiently
+	// confirmed. We can now send back our clean up message, failing the
+	// HTLC on the incoming link.
+	failureMsg := &lnwire.FailPermanentChannelFailure{}
+	err := h.DeliverResolutionMsg(ResolutionMsg{
+		SourceChan: h.ShortChanID,
+		HtlcIndex:  h.htlc.HtlcIndex,
+		Failure:    failureMsg,
+	})
+	if err != nil {
+		return err
+	}
+
+	return h.Checkpoint(h, report)
+}
+
+// checkpointClaim checkpoints the timeout resolver with the reports it needs.
+func (h *htlcTimeoutResolver) checkpointClaim(
+	spendDetail *chainntnfs.SpendDetail) error {
+
+	h.log.Infof("resolving htlc with incoming fail msg, output=%v "+
+		"confirmed in tx=%v", spendDetail.SpentOutPoint,
+		spendDetail.SpenderTxHash)
+
+	// For the direct-timeout spend, we will jump to this checkpoint
+	// without calling `checkpointStageOne`. Thus we need to send the clean
+	// up msg to fail the incoming HTLC.
+	if h.isRemoteCommitOutput() {
+		failureMsg := &lnwire.FailPermanentChannelFailure{}
+		err := h.DeliverResolutionMsg(ResolutionMsg{
+			SourceChan: h.ShortChanID,
+			HtlcIndex:  h.htlc.HtlcIndex,
+			Failure:    failureMsg,
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Send notification.
+	h.ChainArbitratorConfig.HtlcNotifier.NotifyFinalHtlcEvent(
```

Comment:
hmmm...yeah looks like we forgot to call this before, moved to a new commit since it's a bug fix.

Reply:
actually don't think we need this in the timeout resolver since it deals with `update_settle_htlc`. Also not sure if people are still using it, feel like it's hacky way in the [original PR](https://github.com/lightningnetwork/lnd/pull/6517#pullrequestreview-1147152533), maybe we could remove it when the exit settle is fixed.

Reply:
IIRC it was a way to make sure to notify via the normal invoice subscription system that we actually settled an invoice _on disk_.